packages/rum/src/tracer.ts

```diff
@@ -0,0 +1,153 @@
+
+interface BrowserWindow extends Window {
+  ddtrace?: any
+}
+
+export interface Tracer {
+  traceFetch: (context: Partial<FetchContext>) => TraceIdentifier | undefined
+  traceXhr: (context: Partial<XhrContext>, xhr: XMLHttpRequest) => TraceIdentifier | undefined
+}
+
+interface TracingHeaders {
+  [key: string]: string
+}
+
+export function startTracer(configuration: Configuration): Tracer {
+  return {
+    traceFetch: (context) =>
+      onTracingAllowed(configuration, context.url!, (tracingHeaders: TracingHeaders) => {
+        context.init = context.init || {}
+        context.init.headers = context.init.headers || {}
+
+        if (typeof (context.init.headers as Headers).set === 'function') {
+          Object.keys(tracingHeaders).forEach((name) => {
+            ;(context.init!.headers as Headers).set(name, tracingHeaders[name])
+          })
+        } else {
+          assign(context.init.headers, tracingHeaders)
+        }
+      }),
+    traceXhr: (context, xhr) =>
+      onTracingAllowed(configuration, context.url!, (tracingHeaders: TracingHeaders) => {
+        Object.keys(tracingHeaders).forEach((name) => {
+          xhr.setRequestHeader(name, tracingHeaders[name])
+        })
+      }),
+  }
+}
+
+function onTracingAllowed(
+  configuration: Configuration,
+  url: string,
+  inject: (tracingHeaders: TracingHeaders) => void
+): TraceIdentifier | undefined {
+  if (!isTracingSupported() || !configuration.enableTracing || !isAllowedUrl(configuration, url)) {
+    return undefined
+  }
+
+  if (isDdTraceJsActive()) {
+    return getTraceIdFromDdTraceJs()
+  }
+
+  const traceId = new TraceIdentifier()
+  inject(makeTracingHeaders(traceId))
+  return traceId
+}
+
+function isAllowedUrl(configuration: Configuration, url: string) {
+  if (url.indexOf(window.location.origin) === 0) {
+    return true
+  }
+  for (const allowedUrl of configuration.tracingAllowedUrls) {
+    if (allowedUrl.test(url)) {
+      return true
+    }
+  }
+  return false
+}
+
+export function isTracingSupported() {
+  return getCrypto() !== undefined
+}
+
+function getCrypto() {
+  return window.crypto || (window as any).msCrypto
+}
+
+function makeTracingHeaders(traceId: TraceIdentifier): TracingHeaders {
+  return {
+    'x-datadog-origin': 'rum',
+    'x-datadog-parent-id': toDecimalString(traceId),
+    'x-datadog-sampled': '1',
+    'x-datadog-sampling-priority': '1',
+    'x-datadog-trace-id': toDecimalString(traceId),
+  }
+}
+
+export function isDdTraceJsActive() {
+  // tslint:disable-next-line: no-unsafe-any
+  return 'ddtrace' in window && (window as BrowserWindow).ddtrace.tracer.scope().active()
+}
+
+function getTraceIdFromDdTraceJs(): TraceIdentifier | undefined {
+  // tslint:disable-next-line: no-unsafe-any
+  return (window as BrowserWindow).ddtrace.tracer
+    .scope()
+    .active()
+    .context()._traceId // internal trace identifier
+}
+
+/* tslint:disable:no-bitwise */
+export class TraceIdentifier {
```

Comment:
Is it possible to create a unit test to validate those Identifiers?

Reply:
what would you want to validate?

Reply:
The format of the identifier.

Reply:
something like a2525e6?

Reply:
Perfect!